pallets/tendermint-light-client/src/lib.rs

```diff
@@ -0,0 +1,308 @@
+        pub fn submit_finality_proof(
+            origin: OriginFor<T>,
+            untrusted_block: LightBlockStorage,
+        ) -> DispatchResult {
+            ensure_not_halted::<T>()?;
+            let who = ensure_signed(origin)?;
+            let hash = untrusted_block.signed_header.commit.block_id.hash.clone();
+
+            log::debug!(target: "runtime::tendermint-lc", "Verifying light block {:#?}", &hash);
+
+            let options: Options = <LightClientOptions<T>>::get().try_into()?;
+            let verifier = ProdVerifier::default();
+            let most_recent_trusted_block = match <ImportedBlocks<T>>::get(
+                <LastImportedHash<T>>::get(),
+            ) {
+                Some(best_finalized) => best_finalized,
+                None => {
+                    log::error!(
+                        target: "runtime::tendermint-lc",
+                        "Cannot finalize light block {:?} because Light Client is not yet initialized",
+                        &untrusted_block,
+                    );
+                    fail!(<Error<T>>::NotInitialized);
+                }
+            };
+
+            let now = T::TimeProvider::now();
+            let now = Time::from_unix_timestamp(now.as_secs().try_into().unwrap(), 0).unwrap();
+
+            // println!("@ now {:?}", now);
+
+            let verdict = verify_light_block(
+                verifier,
+                untrusted_block.clone(),
+                most_recent_trusted_block,
+                &options,
+                now,
+            );
+
+            match verdict {
+                tendermint_light_client_verifier::Verdict::Success => {
+                    // update storage
+                    insert_light_block::<T>(hash, untrusted_block);
+                    log::info!(target: "runtime::tendermint-lc", "Successfully verified light block {:#?}", &hash);
+                    Self::deposit_event(Event::ImportedLightBlock(who, hash));
+                    Ok(())
+                }
+                tendermint_light_client_verifier::Verdict::NotEnoughTrust(voting_power_tally) => {
+                    log::warn!(target: "runtime::tendermint-lc", "Not enough voting power to accept the light block {:#?}, vote tally  {}", &hash, &voting_power_tally);
+                    fail!(<Error<T>>::NotEnoughTrust)
+                }
+                tendermint_light_client_verifier::Verdict::Invalid(why) => {
+                    log::warn!(target: "runtime::tendermint-lc", "Rejecting invalid light block {:#?} becasue {}", &hash, &why);
+
+                    fail!(<Error<T>>::InvalidBlock)
+                }
+            }
+        }
+
+        /// Halt or resume all light client operations
+        ///
+        /// Can only be called by root
+        #[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+        pub fn set_halted(origin: OriginFor<T>, halted: bool) -> DispatchResult {
```

Comment:
This method should be called by the pallet itself if it detects a fork.

Reply:
added as TODO item